### DISCUSSION TITLE
Minimalist Configuration for a Static Llibrary in C

### DISCUSSION THREAD

**Discussion by saadshams** [microsoft/vcpkg#35234] [Q&A]
Please provide all three minimalist examples of a publisher, VCPKG configuration, and consumer of a static library. For instance, let's say I have a library or publisher.

## Publisher: 

**CMakeLists.txt: A Static Library Project**

```
cmake_minimum_required(VERSION 3.26)
project(simple C)

set(CMAKE_C_STANDARD 11)

add_library(simple library.c)
```

**vcpkg.json**

```
{
  "name" : "simple",
  "version-string" : "1.0.0",
  "description" : "Static Library for C",
  "homepage" : "https://github.com/username/simple",
  "license": "MIT",
  "maintainers" : [ "John Doe" ],
  "dependencies" : [ ]
}
```

## VCPKG

**portfile.cmake**: Please provide

**vcpkg.json**:

```
{
  "name": "simple",
  "version": "1.0.0",
  "maintainers": ["John Doe"],
  "description": "Static Library for C",
  "homepage": "https://github.com/username/simple",
  "license": "MIT",
  "dependencies": [
    {
      "name": "vcpkg-cmake",
      "host": true
    },
    {
      "name": "vcpkg-cmake-config",
      "host": true
    }
  ]
}
```

## Consumer

**Consumer of the Library Project**
**CMakeLists.txt**

```
cmake_minimum_required(VERSION 3.26)
project(consumer C)

set(CMAKE_C_STANDARD 11)

add_executable(consumer main.c)
```

**vcpkg.json**

```
{
  "name": "consumer",
  "version-string": "1.0.0",
  "description": "C",
  "homepage": "",
  "license": "MIT",
  "maintainers": [
    "John Doe"
  ],
  "dependencies": [
    {
      "name": "simple"
    }
  ]
}
```

**Comment by Neumann-A** [comment]
Run https://github.com/friendlyanon/cmake-init

**Reply by saadshams** [reply]
Sorry, @Neumann-A, this is not what I'm looking for; I can't run` cmake-init` for all publisher, vcpkg and consumer of the library.

I need a proper and complete answer using my CMake and vcpkg.json files please.

**Reply by Neumann-A** [reply]
Instead of letting others do your homework how about trying to learn proper CMake yourself? I have given you a rescource which gives you a good basic structure to make consumable CMake packages. 
The problem here is really not vcpkg but limites knowledge how to use cmake.

**Reply by saadshams** [reply]
This is learning; I have attempted the homework. That's why I was able to provide 5 pieces out of 6 and post them here for the missing piece. That's another thing if you don't want to help. 
The problem with cmake-init is that it adds a bunch of pieces that I might not need; hence the bloated configurations, that's why the title asks for a Minimalist example, and this should have been part of **Getting Started** somewhere.

**Reply by Neumann-A** [reply]
There is:

https://learn.microsoft.com/en-us/vcpkg/get_started/get-started-packaging?pivots=shell-cmd

https://github.com/microsoft/vcpkg/blob/master/scripts/templates/portfile.in.cmake

**Reply by Osyotr** [reply]
https://github.com/northwindtraders/beicode looks pretty minimalistic to me (this is from https://devblogs.microsoft.com/cppblog/registries-bring-your-own-libraries-to-vcpkg/ aka https://learn.microsoft.com/en-us/vcpkg/consume/git-registries)

**Comment by saadshams** [comment]
> https://github.com/northwindtraders/beicode looks pretty minimalistic to me

beicode is C++, Is there a miniamalistic C-only library? 

I don't have namespaces and I have a couple C-only function library.

**Reply by saadshams** [reply]
seems I also need the template config.cmake.in for find_package to work for the consumers, where can I find reference and demos to that?

**Reply by dg0yt** [reply]
Reference for CMake packages:  
https://cmake.org/cmake/help/latest/manual/cmake-packages.7.html
You are interested in "Creating packages" of type "Config-file packages".

**Reply by saadshams** [reply]
@dg0yt That is correct. For some reason, `find_package` is unable to find `mylib-onfig.cmake` and `mylib-config-version.cmake` files from the `share` directory.

How do you integrate `find_package` with **VCPKG** so that it searches for configuration files under `vcpkg_installed/x64-osx/share `directory in a platform-agnostic way?

**Reply by dg0yt** [reply]
`share/<PACKAGE_NAME>` is one of the standard paths searched for any prefix known to cmake. And the vcpkg toolchain file adds the installation prefix to `CMAKE_PREFIX_PATH`.

**Reply by saadshams** [reply]
Still my downstream is unable to `find_package`. My upstream [CMakeLists.txt](https://github.com/saadshams/nanojsonc/blob/main/CMakeLists.txt)

I am blocked.

## Consumer 
```
cmake_minimum_required(VERSION 3.26)
project(usagenanojson C)

set(CMAKE_C_STANDARD 11)

add_executable(usagenanojson main.c)

find_package(nanojsonc CONFIG REQUIRED)
target_link_libraries(usagenanojson PRIVATE nanojsonc)
```

![Screen Shot 2023-12-31 at 6 07 45 PM](https://github.com/microsoft/vcpkg/assets/703661/fb5f3d28-b769-417e-8922-fe6a5a138523)

**Reply by Neumann-A** [reply]
did you set `CMAKE_TOOLCHAIN_FILE`?

**Reply by saadshams** [reply]
Thanks for pointing it out. I didn't; is it supposed to be this line before `find_package`?

But I tried with the `CMAKE_TOOLCHAIN_FILE` without any success:
```
cmake_minimum_required(VERSION 3.26)
project(usagenanojson C)
set(CMAKE_C_STANDARD 11)

add_executable(usagenanojson main.c)
set(CMAKE_TOOLCHAIN_FILE "${VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake")
find_package(nanojsonc CONFIG REQUIRED)
target_link_libraries(usagenanojson PRIVATE nanojsonc)
```

I am not sure what I am missing. Do you agree with [Upstream CMakeList.txt](https://github.com/saadshams/nanojsonc/blob/main/CMakeLists.txt)? Maybe there's a glitch there.


**Generated files**
![Screen Shot 2024-01-01 at 8 23 47 AM](https://github.com/microsoft/vcpkg/assets/703661/e06cedfe-9783-404c-af15-0dfdfcee655a)

**Reply by Neumann-A** [reply]
Read the docs at https://learn.microsoft.com/en-us/vcpkg/users/buildsystems/cmake-integration ? ...

**Reply by saadshams** [reply]
> CMake versions older than 3.19 must pass the toolchain file on the configure command line:

Great!. It worked by setting CMake options in the build settings: `-DCMAKE_TOOLCHAIN_FILE=/path/to/vcpkg/scripts/buildsystems/vcpkg.cmake`

![Screen Shot 2024-01-01 at 9 44 57 AM](https://github.com/microsoft/vcpkg/assets/703661/8c034ced-83a8-495e-9354-e60f9e9b70af)


One more, please: the consumer is unable to find headers. Here's my [config.cmake](https://github.com/saadshams/nanojsonc/blob/main/build/config.cmake) just in case.

![Screen Shot 2024-01-01 at 9 46 44 AM](https://github.com/microsoft/vcpkg/assets/703661/cc355072-8f61-4301-bd27-2b3a9bb43556)
![Screen Shot 2024-01-01 at 9 46 59 AM](https://github.com/microsoft/vcpkg/assets/703661/6022846f-130f-44a4-9911-fb0b0a1ba6dc)

**Reply by Neumann-A** [reply]
Installed libraries need to be included with `<>` not `""` otherwise post your `targets.cmake`. The config doesn't help here.

**Reply by saadshams** [reply]
same issue (fatal error: file not found) with both `#include <nanojsonc/parser.h>` and` #include <parser.h>`
![Screen Shot 2024-01-01 at 10 25 46 AM](https://github.com/microsoft/vcpkg/assets/703661/09b8c3f3-f316-4d9a-a97b-4986aac96c6f)

>> otherwise post your targets.cmake

Sorry, I'm not clear on the targets.cmake file, are you referring to the auto-generated file `share/nanojsonc/nanojsonc-targets.cmake`?

**Reply by Neumann-A** [reply]
>  are you referring to the auto-generated file share/nanojsonc/nanojsonc-targets.cmake?

Yes

**Reply by saadshams** [reply]
Sure. **share/nanojsonc/nanojsonc-targets.cmake**

```cmake
# Generated by CMake

if("${CMAKE_MAJOR_VERSION}.${CMAKE_MINOR_VERSION}" LESS 2.8)
   message(FATAL_ERROR "CMake >= 2.8.0 required")
endif()
if(CMAKE_VERSION VERSION_LESS "2.8.3")
   message(FATAL_ERROR "CMake >= 2.8.3 required")
endif()
cmake_policy(PUSH)
cmake_policy(VERSION 2.8.3...3.25)
#----------------------------------------------------------------
# Generated CMake target import file.
#----------------------------------------------------------------

# Commands may need to know the format version.
set(CMAKE_IMPORT_FILE_VERSION 1)

# Protect against multiple inclusion, which would fail when already imported targets are added once more.
set(_cmake_targets_defined "")
set(_cmake_targets_not_defined "")
set(_cmake_expected_targets "")
foreach(_cmake_expected_target IN ITEMS nanojsonc)
  list(APPEND _cmake_expected_targets "${_cmake_expected_target}")
  if(TARGET "${_cmake_expected_target}")
    list(APPEND _cmake_targets_defined "${_cmake_expected_target}")
  else()
    list(APPEND _cmake_targets_not_defined "${_cmake_expected_target}")
  endif()
endforeach()
unset(_cmake_expected_target)
if(_cmake_targets_defined STREQUAL _cmake_expected_targets)
  unset(_cmake_targets_defined)
  unset(_cmake_targets_not_defined)
  unset(_cmake_expected_targets)
  unset(CMAKE_IMPORT_FILE_VERSION)
  cmake_policy(POP)
  return()
endif()
if(NOT _cmake_targets_defined STREQUAL "")
  string(REPLACE ";" ", " _cmake_targets_defined_text "${_cmake_targets_defined}")
  string(REPLACE ";" ", " _cmake_targets_not_defined_text "${_cmake_targets_not_defined}")
  message(FATAL_ERROR "Some (but not all) targets in this export set were already defined.\nTargets Defined: ${_cmake_targets_defined_text}\nTargets not yet defined: ${_cmake_targets_not_defined_text}\n")
endif()
unset(_cmake_targets_defined)
unset(_cmake_targets_not_defined)
unset(_cmake_expected_targets)


# Compute the installation prefix relative to this file.
get_filename_component(_IMPORT_PREFIX "${CMAKE_CURRENT_LIST_FILE}" PATH)
get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)
get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)
if(_IMPORT_PREFIX STREQUAL "/")
  set(_IMPORT_PREFIX "")
endif()

# Create imported target nanojsonc
add_library(nanojsonc STATIC IMPORTED)

set_target_properties(nanojsonc PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
)

# Load information for each installed configuration.
file(GLOB _cmake_config_files "${CMAKE_CURRENT_LIST_DIR}/nanojsonc-targets-*.cmake")
foreach(_cmake_config_file IN LISTS _cmake_config_files)
  include("${_cmake_config_file}")
endforeach()
unset(_cmake_config_file)
unset(_cmake_config_files)

# Cleanup temporary variables.
set(_IMPORT_PREFIX)

# Loop over all imported files and verify that they actually exist
foreach(_cmake_target IN LISTS _cmake_import_check_targets)
  foreach(_cmake_file IN LISTS "_cmake_import_check_files_for_${_cmake_target}")
    if(NOT EXISTS "${_cmake_file}")
      message(FATAL_ERROR "The imported target \"${_cmake_target}\" references the file
   \"${_cmake_file}\"
but this file does not exist.  Possible reasons include:
* The file was deleted, renamed, or moved to another location.
* An install or uninstall procedure did not complete successfully.
* The installation package was faulty and contained
   \"${CMAKE_CURRENT_LIST_FILE}\"
but not all the files it references.
")
    endif()
  endforeach()
  unset(_cmake_file)
  unset("_cmake_import_check_files_for_${_cmake_target}")
endforeach()
unset(_cmake_target)
unset(_cmake_import_check_targets)

# This file does not depend on other imported targets which have
# been exported from the same project but in a separate export set.

# Commands beyond this point should not need to know the version.
set(CMAKE_IMPORT_FILE_VERSION)
cmake_policy(POP)


```

**Reply by Neumann-A** [reply]
```
# Create imported target nanojsonc
add_library(nanojsonc STATIC IMPORTED)

set_target_properties(nanojsonc PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
)
```
looks good.

```
# Compute the installation prefix relative to this file.
get_filename_component(_IMPORT_PREFIX "${CMAKE_CURRENT_LIST_FILE}" PATH)
get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)
get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)
if(_IMPORT_PREFIX STREQUAL "/")
  set(_IMPORT_PREFIX "")
endif()
```
Looks also good (since the file is in `share/nanojsonc/nanojsonc-targets.cmake`, so import_prefix should be pointing to the root of the installed folder.) Wait but you also mentioned  `share/nanojsonc-targets.cmake` ? So where is the file now?
The last thing which could happen is that in `target_link_libraries(usagenanojson PRIVATE nanojsonc)` `nanojsonc` is not the target for some reason. This is why you would normally namespace the target to avoid this since only targets are allowed to have `::` in them.

**Reply by saadshams** [reply]
sorry, I fixed the path in my last post, it is `share/nanojsonc/nanojsonc-targets.cmake`

>> This is why you would normally namespace the target to avoid this since only targets are allowed to have :: in them.

If I'm right, are you suggesting changing the line 40 from upstream [CMakeLists.txt](https://github.com/saadshams/nanojsonc/blob/main/CMakeLists.txt#L40) to

```
# Install Export Set
install(EXPORT ${PROJECT_NAME}-targets
        NAMESPACE nanojsonc::
        DESTINATION ${CMAKE_INSTALL_DATADIR}/${PROJECT_NAME})
```

**Reply by saadshams** [reply]
This didn't work either. Blocked :(
`target_link_libraries(usagenanojson PRIVATE nanojsonc::nanojsonc)`

**Reply by Neumann-A** [reply]
So still the include error ?

**Reply by Neumann-A** [reply]
Are you invoking the build manually instead of running `cmake --build` or through the geneator?

**Reply by saadshams** [reply]
Good catch! I had to fix that auto-run button in my IDE and I was able to link.
I just have never seen undefined symbol errors before; maybe it's on my end. Let me know please if that's the case. 

Nevertheless, I very much appreciate your support, and a happy new year to you and your colleagues.


![Screen Shot 2024-01-01 at 3 35 46 PM](https://github.com/microsoft/vcpkg/assets/703661/db81414e-fbd2-48d6-befe-88695c967dff)

**Reply by saadshams** [reply]
1. Turns out to be the sanitizer issue. fixed:
```
if (CMAKE_VERSION VERSION_GREATER_EQUAL 3.1 AND NOT WIN32) # sanitizer
    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fsanitize=address,undefined -fno-sanitize-recover=all")
    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fsanitize=address,undefined")
endif ()
```

2. nano zone abandoned error 

`usagenanojson(88210,0x102c85600) malloc: nano zone abandoned due to inability to preallocate reserved vm space.`

Fixed by setting env MallocNanoZone=0. [Reference](https://stackoverflow.com/questions/64126942/malloc-nano-zone-abandoned-due-to-inability-to-preallocate-reserved-vm-space)

I'm not sure if that's a good practice.

**Reply by Neumann-A** [reply]
Those are on you. If you build your lib as a static lib with asan you need to build everything downstream with asan or at least link the asan libs.

**Reply by saadshams** [reply]
Thanks. I'll get back soon on the pull request to push this library and the others since I have a working template now. All is good now.

Thanks again and a happy new year to everyone on the team.

**Reply by saadshams** [reply]
@Neumann-A 

The library is ready to go. Kindly push the library to the registry. Let me know, please.
